core/templates/pages/exploration-player-page/layout-directives/content-language-selector.component.spec.ts

```diff
@@ -117,16 +125,24 @@ describe('Content language selector component', () => {
     fixture.detectChanges();
   }));
 
-  it('should correctly initialize selectedLanguageCode and ' +
-     'languagesInExploration', () => {
+  it('should correctly initialize selectedLanguageCode, ' +
+    'newLanguageCode and languagesInExploration', () => {
     expect(component.selectedLanguageCode).toBe('fr');
     expect(component.languageOptions).toEqual([
-      {value: 'fr', displayed: 'français (French)'},
-      {value: 'zh', displayed: '中文 (Chinese)'},
-      {value: 'en', displayed: 'English'}
+      { value: 'fr', displayed: 'français (French)' },
+      { value: 'zh', displayed: '中文 (Chinese)' },
+      { value: 'en', displayed: 'English' }
     ]);
   });
 
+  it('should correcly initialize newLanguageCode', () => {
+    let mockWindowRef = new MockWindowRef();
+    let url = new URL(mockWindowRef.nativeWindow.location.href);
+    expect(url.searchParams.get('initialContentLanguageCode')).toBe('fr');
+    component.ngOnInit();
```

Comment:
Thanks! @gp201 When i did that the value is still comming out to be 'fr' even if i change the url initial language code it is like we are not linked to windowRef we always get the same value.

Any suggesion ? I am really sorry to bother you but I dont have any clue how to move further.

Reply:
@Priyansh61 you seem to have forgotten to direct the component to use the class that was defined.
https://github.com/oppia/oppia/blob/5d951d8e21e2f7872dcb26e8618952412efe4312/core/templates/services/user.service.spec.ts#L57

Take a look at this file above on how to mock `WindowRef`.

Reply:
Thanks @gp201 I tried to use the windowRef provider and MockWindow Ref as useClass the test for new language code passes succesfully for the first time :sweat_smile:  the url changes and the newlanguageCode initialised itself according to the url.

Then the problem the remaining 3 cases suddenly fails and gives me the following error.

````console
Content language selector component should correctly select an option when refresh is not needed FAILED
        TypeError: Cannot read properties of undefined (reading 'split')
            at ContextService.getPageContext (core/templates/combined-tests.spec.js:213435:57336)
            at ContextService.isInExplorationEditorPage (core/templates/combined-tests.spec.js:213435:71135)
            at ContentLanguageSelectorComponent.onSelectLanguage (core/templates/combined-tests.spec.js:148323:25192)
            at UserContext.<anonymous> (core/templates/combined-tests.spec.js:148226:19)
            at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:733524:30)
            at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:733009:43)
            at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:733523:36)
            at ./node_modules/zone.js/dist/zone.js.Zone.run (core/templates/combined-tests.spec.js:733281:47)
            at runInTestZone (core/templates/combined-tests.spec.js:732564:38)
            at UserContext.<anonymous> (core/templates/combined-tests.spec.js:732579:24)
````


Can you please help me with this ? The only change I made is 

```typescript
{
        provide: WindowRef,
       useClass: MockWindowRef
},
```

The above change makes the new test case pass but fails all the other tests

Reply:
The error given below:
```console
TypeError: Cannot read properties of undefined (reading 'split')
```

tells that there is an undefined value. Can you check if a variable that is used in other tests hasn't been defined?

Reply:
Thank you very much, @gp201, for guiding me. Actually, one of the tests required the windowRef's pathname attribute as well. I included it now, and all of the tests are finally passing. I also verified for both scenarios when searchPram is present and when it is not there, and if present, it returns en and otherwise fr.